persistent/Database/Persist/TH.hs

```diff
@@ -1780,6 +1776,18 @@ mkKeyTypeDec mps entDef = do
     supplement :: [Name] -> [Name]
     supplement names = names <> (filter (`notElem` names) $ mpsDeriveInstances mps)
 
+    -- Always use StockStrategy for Show/Read.
+    -- This means e.g. (FooKey 1) shows as ("FooKey 1"), rather than just "1".
+    -- This is much better for debugging/logging purposes:
+    -- cf. https://github.com/yesodweb/persistent/issues/1104
+    decideStrategy :: Name -> DerivStrategy
+    decideStrategy typeclass
+        | typeclass `elem` [''Show, ''Read] = StockStrategy
+        | typeclass == ''PathMultiPiece =
+            ViaStrategy $ ConT ''ViaPersistEntity `AppT` recordType
```

Comment:
Apologies, I should have been more clear - I'm not going to accept a PR that has a special-case for how an instance is derived. 

```suggestion
```

Reply:
OK, so regarding your first comment,

> I'm OK with introducing features to make manual instances easier - ie exporting the function defaults that are based on `PersistEntity a => ...`, and a `newtype` for `DerivingVia`.

should I revert my changes to `Database.Persist.TH` and remove `CompositeKeyPathMultiPieceSpec`? That would leave `ViaPersistEntity` and its `PathMultiPiece` instance as the additions from this PR. If those don't seem useful on their own, I'm fine with closing the PR.

I'll keep the rest on my fork as automatic `PathMultiPiece` derivation is helpful for my specific use case, although I don't think it would be used often if it was added here (no issues have been raised to request it).

Reply:
Yes, that works.

I need to document more about how to hook into the derivation stuff without modifying this function here.

The way that I would approach this is with an additional function that operates on the `[UnboundEntityDef]` from a thing. So, you have,

```haskell
$(share [mkPersist sqlSettings, derivePathMultiPiece] ...)
```

Where `derivePathMultiPiece :: [UnboundEntityDef] -> Q [Dec]`, and you can do something like:

```haskell
derivePathMultiPiece = 
  concat <$> traverse derivePathMultiPieceFor
  where 
    derivePathMultiPieceFor unboundEntityDef = 
      let keyTypeName = ...
      [d| deriving via (ViaPersistentEntity $(keyTypeName)) instance PathMultiPiece $(keyTypeName) |]
```

This gives you the same automatic instance generation, without modifying the upstream code.